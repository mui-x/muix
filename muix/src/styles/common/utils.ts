export function deepMerge(target, ...sources) {
  if (!sources.length) return target

  const source = sources.shift()
  if (isObject(target) && isObject(source)) {
    for (const key in source) {
      if (isObject(source[key])) {
        if (!target[key]) target[key] = {}
        deepMerge(target[key], source[key])
      } else 
        target[key] = source[key]
    }
  } else
    throw 'Cannot merge object and non object props'

  return deepMerge(target, ...sources);
}

const isObject = item => item && typeof item === 'object' && !Array.isArray(item)
