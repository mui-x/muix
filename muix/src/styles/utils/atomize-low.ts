import { wrapPseudoPrefixes } from './wrap-pseudo-prefixes'
import { TEngine } from 'reactxx-typings'

// platform dependent 
import { platform } from 'reactxx-styles'

//*******************************************************
//        adjustAtomizedLow

// linearize and atomize "ruleset"
// result is added to 'atomizedVariants'
export const adjustAtomizedLow = (
    ruleset: TEngine.Rulesets,
    atomizedVariants: TEngine.QueryableItems, path: string,
    pseudoPrefixes: string[], conditions: TEngine.Conditions
) => {
    processTree(ruleset, atomizedVariants, path, pseudoPrefixes, conditions)
}

export function isToAtomizeArray(obj): obj is TEngine.Ruleset[] { return obj && !(obj as TEngine.Queryables).$r$ && Array.isArray(obj) }
export function isAtomized(obj): obj is TEngine.Queryables { return !obj || (obj as TEngine.Queryables).$r$ }
export function isDeferred(obj): obj is TEngine.Deferred { return obj && (obj as TEngine.Deferred).$d$ }
export function isTemporary(obj): obj is TEngine.TempProc { return obj && (obj as TEngine.TempProc).$t$ }
export function isToAtomize(obj): obj is TEngine.ToAtomize { return obj && !(obj as TEngine.Queryables).$r$ && !(obj as TEngine.Deferred).$d$ && !(obj as TEngine.TempProc).$t$ }

//*******************************************************
//        makeTemporary
export const makeTemporary = (proc: TEngine.TempProc) => {
    proc.$t$ = true
    return proc as any
}

//*******************************************************
//        processTree
export const processTree = (
    value: TEngine.Rulesets, atomizedVariants: TEngine.QueryableItems,
    path: string, pseudoPrefixes: string[], conditions: TEngine.Conditions
) => {

    const indexToPath = (idx: number) => items.length <= 1 ? '' : `[${idx}]`

    const items = isToAtomizeArray(value) ? value : [value]
    items.forEach((it, idx) => {
        if (!it) return
        if (isTemporary(it)) {
            it(atomizedVariants, path + indexToPath(idx), pseudoPrefixes, conditions)
        } else if (isDeferred(it)) {
            atomizedVariants.push(it)
        } else if (isAtomized(it)) {
            if (pseudoPrefixes && pseudoPrefixes.length > 0)
                throw 'Web pseudo properties cannot contain already atomized value'
            for (const itt of it) {
                if (isDeferred(itt))
                    throw 'isDeferred(itt)'
                if (!conditions || conditions.length === 0) //... and no conditions => done
                    pushToAtomizedVariants(atomizedVariants, itt, null)
                else { // ... and conditions => merge conditions
                    const ittCopy = Array.isArray(itt) ? [...itt] : { ...itt }
                    pushToAtomizedVariants(atomizedVariants, ittCopy,
                        itt.conditions ? [...conditions, ...itt.conditions] : conditions)
                }
            }
        } else { // it: toAtomize
            // if (window.__TRACE__)
            //     it = deepClone(it)
            processPseudosAndAtomize(
                it,
                atomizedVariants, path + indexToPath(idx),
                pseudoPrefixes, conditions
            )
        }
    })
}

//*********************************************************
//  PRIVATE
//*********************************************************

//*******************************************************
//        processAllInnerVariants

// recursivelly process web pseudos (:hover, :active) and its queryable parts

const processPseudosAndAtomize = (
    ruleset: TEngine.ToAtomize,
    atomizedVariants: TEngine.QueryableItems, path: string, pseudoPrefixes: string[], conditions: TEngine.Conditions,
) => {

    const inner: TEngine.QueryableItems = []

    const toAtomizePure = {}
    let empty = true
    // process and remove conditional parts of self and web pseudo (e.g. :hover etc.)
    // push them to 'inner'
    for (const p in ruleset) {
        if (p.charAt(0) === '$') continue
        const value = ruleset[p] as TEngine.Rulesets
        if (typeof value === 'object' || isTemporary(value)) { // temporary, deffered, toAtomize, atomized
            processTree(value, inner, `${path}/${p}`, [...pseudoPrefixes, p], conditions)
        } else {
            empty = false
            toAtomizePure[p] = value
        }
    }

    // atomize pure ruleset tree with pseudo (and without temporary, deffer's etc.),
    // then push to result
    if (!empty)
        atomizePure(atomizedVariants, wrapPseudoPrefixes(toAtomizePure, pseudoPrefixes), conditions, path)

    // push 'inner' to result AFTER toAtomizePure
    if (inner.length > 0)
        Array.prototype.push.apply(atomizedVariants, inner)

}

// *********** utils

const pushToAtomizedVariants = (
    atomizedVariants: TEngine.QueryableItems,
    variant: TEngine.Queryable, conditions: TEngine.Conditions
) => {
    if (!variant) return
    if (conditions && conditions.length > 0) variant.conditions = conditions
    atomizedVariants.push(variant)
}

const atomizePure = (
    atomizedVariants: TEngine.QueryableItems, pureRuleset: TEngine.ToAtomize,
    conditions: TEngine.Conditions, path: string
) => {
    if (!pureRuleset) return
    if (isDeferred(pureRuleset)) {
        throw 'something wrong'
        atomizedVariants.push(pureRuleset)
        return
    }
    // platform specific atomize: e.g. for WEB convert ruleset to list of atomized CSS classes
    const variant = platform.toPlatformAtomizeRuleset(pureRuleset, path)
    pushToAtomizedVariants(atomizedVariants, variant, conditions)
}

export function deepClone(obj) {
    if (
        obj === undefined || obj === null || typeof obj === 'string' || typeof obj === 'undefined' ||
        typeof obj === 'function' || typeof obj === 'number' || typeof obj === 'boolean'
    ) return obj

    if (Array.isArray(obj)) {
        const res = obj.map(o => deepClone(o))
        if (obj['$r$']) res['$r$'] = true
        return res
    }

    if (typeof obj === 'object') {
        const res = {}
        for (const p in obj) res[p] = deepClone(obj[p])
        return res
    }

    throw 'something wrong'
}