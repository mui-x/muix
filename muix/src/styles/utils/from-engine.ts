import { TEngine, TTyped } from 'reactxx-typings';

export const TAsEngineClassName = <R extends TTyped.Shape>(r: TEngine.RulesetOrCreator) => r as TTyped.ClassNameSimple<R>
export const TAsTypedClassName = (r: TTyped.RulesetOrCreator) => r as TEngine.RulesetOrCreator

export const TAsEngineSheet = <R extends TTyped.Shape>(r: TEngine.SheetOrCreator) => r as any as TTyped.SheetSimple<R>
export const TAsTypedSheet = (r: TTyped.PartialSheetOrCreator) => r as any as TEngine.Sheet

export const TAsEngineStyle = <R extends TTyped.Shape>(r: TEngine.StyleOrCreator) => r as any as TTyped.StyleSimple<R>
export const TAsTypedStyle = (r: TTyped.StyleOrCreator) => r as any as TEngine.Style

// export function isReactXXComponent(obj): obj is TEngine.IsReactXXComponent {
//     return (obj as TEngine.IsReactXXComponent).$c$
// }