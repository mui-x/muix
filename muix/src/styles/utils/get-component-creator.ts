//import React from 'react'
import { TTyped, TComponents } from 'reactxx-typings';
import { useStyles } from 'reactxx-styles'

export const getComponentCreator = <R extends TTyped.Shape>(
    getComp: TComponents.GetComponent<R>, authorDisplayName?: string, authorConfig?: TComponents.AuthorConfig<R>, par?
) => (
    userDisplayName?: string, userConfig?: TComponents.UserConfig<R>
) => {
        const config: TComponents.Config<R> = {
            displayName: userDisplayName || authorDisplayName
            //componentId: 0 - platform is not initialized yet
        }
        const cfg = userConfig && authorConfig ? Object.assign({}, authorConfig, userConfig) : userConfig || authorConfig
        if (cfg) Object.assign(config, cfg)
        const _useStyles = (props: TComponents.Props<R>) => useStyles<R>(props, config)
        const Comp = getComp(_useStyles, par)
        Comp.displayName = config.displayName
        return Comp
    }

export const getComponentCreatorUntyped = (authorDisplayName, authorConfig, getComp, par?) =>
    getComponentCreator(getComp, authorDisplayName, authorConfig, par) as any

