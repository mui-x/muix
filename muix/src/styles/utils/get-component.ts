import warning from 'warning'

import { TTyped, TComponents } from 'reactxx-typings';

export const getComponent = <R extends TTyped.Shape>(
    getComp: TComponents.GetComponent<R>, configss?: TComponents.ComponentConfigss<R>
) => {
    const config = mergeConfig(configss)
    config.getComponent = getComp

    const Comp = config.Comp = getComp(config)
    Comp.displayName = config.displayName
    Comp.config = config

    return Comp
}

export const getComponentUntyped = getComponent as any

const mergeConfig = <R extends TTyped.Shape>(inputss: TComponents.ComponentConfigss<R>, config?: TComponents.Config<R>) => {
    if (!config) config = {}
    if (!inputss) return config
    const doMerge = (input: TComponents.ComponentConfig) => {
        if (!input) return
        for (const p in input) {
            if (p.charAt(0) !== '$')
                config[p] = input[p]
            else {
                if (!config[p]) config[p] = []
                config[p].push(input[p])
            }
        }
    }
    if (Array.isArray(inputss))
        for (const inputs of inputss)
            if (Array.isArray(inputs))
                for (const input of inputs)
                    doMerge(input)
            else
                doMerge(inputs)
    else
        doMerge(inputss)
    return config
}
