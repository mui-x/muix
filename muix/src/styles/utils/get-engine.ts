import { TTyped } from 'reactxx-typings';
import { 
    THEMED, QUERY, COMPILE, STYLE, WIDTH, WEB, NATIVE, HOT, IF, IFELSE, COMPONENT
} from '../queryable/index'

export const getTypedEngine = <R extends TTyped.Shape>() => engine as TTyped.TypedEngine<R>

const engine = {
    WEB,
    NATIVE,
    STYLE,
    COMPILE,
    HOT,
    IF,
    IFELSE,
    WIDTH,
    THEMED,
    QUERY,
    COMPONENT,
}

