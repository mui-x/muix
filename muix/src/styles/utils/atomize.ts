import warning from 'warning'

import { TEngine } from 'reactxx-typings'

import { adjustAtomizedLow, isDeferred, deepClone } from './atomize-low'
import { createWithTheme } from './create-with-theme'
import { isStyledComponentTemp } from '../hooks/use-styled-components'

// mutable
export const atomizeSheet = (sheet: TEngine.SheetOrCreator, theme?, path: string = 'sheet') => {
    if (!sheet) return null
    if (window.__TRACE__)
        sheet = deepClone(sheet)
    const sh: TEngine.Sheet = createWithTheme(sheet, theme)
    for (const p in sh) {
        const value = sh[p]
        const subPath = (path ? path + '.' : '') + p
        if (isStyledComponentTemp(value)) {
            warning(p.charAt(0)===p.charAt(0).toUpperCase(), 'Component definition name must start with uppercase letter')
            sh[p] = value(sh, p, subPath)
        } else {
            const at = atomizeRuleset(sh[p], theme, subPath)
            if (at) sh[p] = at
            else delete sh[p]
        }
    }
    return sh
}

// mutable
export const atomizeRuleset = (
    ruleset: TEngine.RulesetOrCreator, theme?, path: string = ''
) => {
    if (!ruleset) return null

    const rs = createWithTheme(ruleset, theme)

    if (!rs) return null

    const list = wrapQueryables([])
    adjustAtomizedLow(rs, list, path, [], [])

    return list.length === 0 ? null : list
}

export const wrapAtomicLow = (atomicLow: TEngine.AtomicLow) => {
    const res = [
        atomicLow
    ] as TEngine.Queryables
    res.$r$ = true
    return res
    // (queryables as TEngine.Queryables).$r$ = true
    // if (window.__TRACE__)
    //     queryables['toJSON'] = toJSON.bind(queryables)
    // return queryables as TEngine.Queryables
}


export const wrapQueryables = (queryables: TEngine.QueryableItems) => {
    (queryables as TEngine.Queryables).$r$ = true
    if (window.__TRACE__)
        queryables['toJSON'] = toJSON.bind(queryables)
    return queryables as TEngine.Queryables
}
function toJSON() {
    return (this as TEngine.Queryables).map(v => {
        if (isDeferred(v)) return 'DEFFERED'
        return v
    })
}

// mutable
export const atomizeStyle = (style: TEngine.Style, theme?) => {
    if (!style) return null

    style = createWithTheme(style, theme)

    const { $web, $native } = style

    if ($web || $native) style = { ...style }

    if (window.isWeb && $web)
        Object.assign(style, $web)
    else if (!window.isWeb && $native)
        Object.assign(style, $native)

    delete style.$web
    delete style.$native

    return style
}
