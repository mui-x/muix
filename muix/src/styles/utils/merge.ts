import { TEngine, TTyped, TComponents } from 'reactxx-typings';
import { deepMerge, deepMerges } from './deep-merge'
import { isAtomized, isDeferred } from './atomize-low';
import { wrapQueryables } from './atomize';
import { isStyledComponentData, isStyledComponent } from '../hooks/use-styled-components'
import warning = require('warning');

export const mergePropsCode = (propsCode: TTyped.PropsCode | any, props: TComponents.Props[]) => {
    if (!props || props.length === 0) return
    let rootWebProps, rootNativeProps, rootProps
    for (const p of props) {
        if (!p) continue
        Object.assign(propsCode, p)
        // merge child component root props
        const { $rootWebProps, $rootNativeProps, $rootProps } = p
        window.isWeb && $rootWebProps && Object.assign(rootWebProps || (rootWebProps = {}), $rootWebProps)
        !window.isWeb && $rootNativeProps && Object.assign(rootNativeProps || (rootNativeProps = {}), $rootNativeProps)
        $rootProps && Object.assign(rootProps || (rootProps = {}), $rootProps)
    }
    if (rootWebProps) propsCode.$rootWebProps = rootWebProps
    if (rootNativeProps) propsCode.$rootNativeProps = rootNativeProps
    if (rootProps) propsCode.$rootProps = rootProps
}

// immutable
export const mergeRulesets = (sources: TEngine.Queryables[]) => {
    if (!sources || sources.length === 0) return null
    if (sources.length === 1) return sources[0] || null

    let res: TEngine.Queryables = null
    let first = true
    let noAtomized = false
    for (const src of sources) {
        if (!src) continue
        if (isAtomized(src)) {
            res = noAtomized || first ? src : [...res, ...src]
            first = noAtomized = false
            continue
        }
        if (isStyledComponentData(src) || isDeferred(src) || isStyledComponent(src)) {
            res = src
            noAtomized = true
            continue
        }
        throw 'All rulesets must be atomized first for mergeRulesets'
    }
    return noAtomized ? res : wrapQueryables(res)
}

// immutable
export const mergeSheets = (sources: TEngine.Sheet[]) => {
    if (!sources || sources.length === 0) return null
    if (sources.length===1) return sources[0] || null

    const ruleLists: Record<string, Array<any>> = {}
    for (const src of sources) {
        if (!src) continue
        for (const p in src) {
            const value = src[p]
            if (!value) continue
            const arr = ruleLists[p] || (ruleLists[p] = [])
            arr.push(value)
        }
    }
    const res = {}
    for (const p in ruleLists)
        res[p] = mergeRulesets(ruleLists[p])
    return res as TEngine.Sheet
}

export const mergeDeep = (sources: {}[]) => {
    if (!sources || sources.length === 0) return null
    let count = 0
    let res = null
    for (const src of sources) {
        if (!src) continue
        switch (count) {
            case 0: res = src; count++; break
            case 1: res = deepMerges({}, [res, src]); count++; break
            default: res = deepMerge(res, src); count++; break
        }
    }
    return res
}

