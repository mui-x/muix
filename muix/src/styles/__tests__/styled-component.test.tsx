import React from 'react'

import { fireEvent } from 'react-testing-library'

import { initPlatform, render, cleanup } from "./init-platform.t"
import { TComponents, $W, $T, $V, $I, V, T, I, TTyped } from 'reactxx-typings'

import { getTypedEngine, useStyles, getComponent } from 'reactxx-styles'
import { mergePropsCode } from '../utils/merge';

//*** INNER COMPONENT

interface iShapeLow extends TTyped.ShapeAncestor {
  props: { enabled?: boolean }
}

const i = getTypedEngine<iShapeLow>()

const isheet = {
  root: i.STYLE<T>(
    { backgroundColor: 'yellow' },
    i.IFELSE<T>(p => p.enabled, { color: 'red' }, { color: 'gray' })
  )
}

interface iShape extends iShapeLow {
  sheet: typeof isheet
}

const iconfig: TComponents.ComponentConfig<iShape> = {
  displayName: 'inner-component',
  $sheet: isheet
}

const getInner: TComponents.GetComponent<iShape> = config => props => {
  const { getRootWebStyleProps, propsCode: { enabled, children } } = useStyles<iShape>(props, config)
  return <span {...getRootWebStyleProps()}>
    INNER  {enabled ? 'ENABLED' : 'DISABLED'}: {children}
  </span>
}

const Inner = getComponent(getInner, iconfig)

//*** OUTER COMPONENT

interface ShapeLow extends TTyped.ShapeAncestor {
  props: { openedInit?: boolean }
  sheetQuery: { opened?: boolean }
}

const { COMPONENT, IF, IFELSE, WEB, NATIVE, STYLE, COMPILE
} = getTypedEngine<ShapeLow>()

const sheet = {
  root: STYLE<V>(
    IFELSE<V>(p => p.opened, { padding: 0 }, { padding: 10 })
  ),
  Left: COMPONENT(Inner, {
    root: STYLE<T>(
      {
        borderColor: 'red'
      },
      IFELSE<T>(p => p.opened, { fontWeight: '500' }, { fontWeight: '300' })
    )
  }),
  Right: COMPONENT(Inner, {
    root: STYLE<T>({
      borderColor: 'green'
    })
  })

}

interface Shape extends ShapeLow {
  sheet: typeof sheet
}

const config: TComponents.ComponentConfig<Shape> = {
  displayName: 'outer-component',
  $sheet: sheet
}

const p: TTyped.getProps<Shape> = {}

const getOuter: TComponents.GetComponent<Shape> = config => props => {
  const { getRootWebStyleProps, classes, propsCode, propsCode: { openedInit, children } } = useStyles<Shape>(props, config)

  const [openedState, setOpenedState] = React.useState(openedInit)
  // set state to sheet query
  propsCode.opened = openedState

  return <div data-testid='toggle' {...getRootWebStyleProps()} onClick={() => setOpenedState(!openedState)}>
    OUTER {openedState ? 'OPENED' : 'CLOSED'}: {children}
    <classes.Left>LEFT</classes.Left>
    <classes.Right enabled>RIGHT</classes.Right>
  </div>
}

const Outer = getComponent(getOuter, config)

describe("STYLED COMPONENT", () => {

  const doTest = (isWeb: boolean) => {
    beforeEach(() => initPlatform(isWeb))
    afterEach(cleanup)


    it("01: Outer", () => {
      const wrapper = render(<Outer />)
      expect(wrapper.container).toMatchSnapshot()
    })

    it("02: Outer opened", () => {
      const wrapper = render(<Outer openedInit />)
      expect(wrapper.container).toMatchSnapshot()
    })

    it("03: Outer both", () => {
      const wrapper = render(<div>
        <Outer openedInit />
        <Outer openedInit={false} />
      </div>)
      expect(wrapper.container).toMatchSnapshot()
    })

    it("04: change open with click", () => {
      const {getByTestId, container} = render(<div>
        <Outer/>
      </div>)
      const toggle = getByTestId('toggle')
      expect(container).toMatchSnapshot()
      fireEvent.click(toggle)
      expect(container).toMatchSnapshot()
    })

  }

  describe("## NATIVE ##", () => doTest(false))
  describe("## WEB ##", () => doTest(true))

})
