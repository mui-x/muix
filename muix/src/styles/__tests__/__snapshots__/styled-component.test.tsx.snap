// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`STYLED COMPONENT ## NATIVE ## 01: Outer 1`] = `
<div>
  <div
    data-testid="toggle"
    data-trace="
padding @outer-component.sheet.root/ifelse-false"
    style="padding: 10px;"
  >
    OUTER 
    CLOSED
    : 
    <span
      data-trace="
backgroundColor @inner-component.sheet.root[0]
color @inner-component.sheet.root[1]/ifelse-false
borderColor @outer-component.sheet.Left/inner-component.root[0]
fontWeight @outer-component.sheet.Left/inner-component.root[1]/ifelse-false"
      style="background-color: yellow; color: gray; border-color: red; font-weight: 300;"
    >
      INNER  
      DISABLED
      : 
      LEFT
    </span>
    <span
      data-trace="
backgroundColor @inner-component.sheet.root[0]
color @inner-component.sheet.root[1]/ifelse-true
borderColor @outer-component.sheet.Right/inner-component.root"
      style="background-color: yellow; color: red; border-color: green;"
    >
      INNER  
      ENABLED
      : 
      RIGHT
    </span>
  </div>
</div>
`;

exports[`STYLED COMPONENT ## NATIVE ## 02: Outer opened 1`] = `
<div>
  <div
    data-testid="toggle"
    data-trace="
padding @outer-component.sheet.root/ifelse-true"
    style="padding: 0px;"
  >
    OUTER 
    OPENED
    : 
    <span
      data-trace="
backgroundColor @inner-component.sheet.root[0]
color @inner-component.sheet.root[1]/ifelse-false
borderColor @outer-component.sheet.Left/inner-component.root[0]
fontWeight @outer-component.sheet.Left/inner-component.root[1]/ifelse-true"
      style="background-color: yellow; color: gray; border-color: red; font-weight: 500;"
    >
      INNER  
      DISABLED
      : 
      LEFT
    </span>
    <span
      data-trace="
backgroundColor @inner-component.sheet.root[0]
color @inner-component.sheet.root[1]/ifelse-true
borderColor @outer-component.sheet.Right/inner-component.root"
      style="background-color: yellow; color: red; border-color: green;"
    >
      INNER  
      ENABLED
      : 
      RIGHT
    </span>
  </div>
</div>
`;

exports[`STYLED COMPONENT ## NATIVE ## 03: Outer both 1`] = `
<div>
  <div>
    <div
      data-testid="toggle"
      data-trace="
padding @outer-component.sheet.root/ifelse-true"
      style="padding: 0px;"
    >
      OUTER 
      OPENED
      : 
      <span
        data-trace="
backgroundColor @inner-component.sheet.root[0]
color @inner-component.sheet.root[1]/ifelse-false
borderColor @outer-component.sheet.Left/inner-component.root[0]
fontWeight @outer-component.sheet.Left/inner-component.root[1]/ifelse-true"
        style="background-color: yellow; color: gray; border-color: red; font-weight: 500;"
      >
        INNER  
        DISABLED
        : 
        LEFT
      </span>
      <span
        data-trace="
backgroundColor @inner-component.sheet.root[0]
color @inner-component.sheet.root[1]/ifelse-true
borderColor @outer-component.sheet.Right/inner-component.root"
        style="background-color: yellow; color: red; border-color: green;"
      >
        INNER  
        ENABLED
        : 
        RIGHT
      </span>
    </div>
    <div
      data-testid="toggle"
      data-trace="
padding @outer-component.sheet.root/ifelse-false"
      style="padding: 10px;"
    >
      OUTER 
      CLOSED
      : 
      <span
        data-trace="
backgroundColor @inner-component.sheet.root[0]
color @inner-component.sheet.root[1]/ifelse-false
borderColor @outer-component.sheet.Left/inner-component.root[0]
fontWeight @outer-component.sheet.Left/inner-component.root[1]/ifelse-false"
        style="background-color: yellow; color: gray; border-color: red; font-weight: 300;"
      >
        INNER  
        DISABLED
        : 
        LEFT
      </span>
      <span
        data-trace="
backgroundColor @inner-component.sheet.root[0]
color @inner-component.sheet.root[1]/ifelse-true
borderColor @outer-component.sheet.Right/inner-component.root"
        style="background-color: yellow; color: red; border-color: green;"
      >
        INNER  
        ENABLED
        : 
        RIGHT
      </span>
    </div>
  </div>
</div>
`;

exports[`STYLED COMPONENT ## NATIVE ## 04: change open with click 1`] = `
<div>
  <div>
    <div
      data-testid="toggle"
      data-trace="
padding @outer-component.sheet.root/ifelse-false"
      style="padding: 10px;"
    >
      OUTER 
      CLOSED
      : 
      <span
        data-trace="
backgroundColor @inner-component.sheet.root[0]
color @inner-component.sheet.root[1]/ifelse-false
borderColor @outer-component.sheet.Left/inner-component.root[0]
fontWeight @outer-component.sheet.Left/inner-component.root[1]/ifelse-false"
        style="background-color: yellow; color: gray; border-color: red; font-weight: 300;"
      >
        INNER  
        DISABLED
        : 
        LEFT
      </span>
      <span
        data-trace="
backgroundColor @inner-component.sheet.root[0]
color @inner-component.sheet.root[1]/ifelse-true
borderColor @outer-component.sheet.Right/inner-component.root"
        style="background-color: yellow; color: red; border-color: green;"
      >
        INNER  
        ENABLED
        : 
        RIGHT
      </span>
    </div>
  </div>
</div>
`;

exports[`STYLED COMPONENT ## NATIVE ## 04: change open with click 2`] = `
<div>
  <div>
    <div
      data-testid="toggle"
      data-trace="
padding @outer-component.sheet.root/ifelse-true"
      style="padding: 0px;"
    >
      OUTER 
      OPENED
      : 
      <span
        data-trace="
backgroundColor @inner-component.sheet.root[0]
color @inner-component.sheet.root[1]/ifelse-false
borderColor @outer-component.sheet.Left/inner-component.root[0]
fontWeight @outer-component.sheet.Left/inner-component.root[1]/ifelse-true"
        style="background-color: yellow; color: gray; border-color: red; font-weight: 500;"
      >
        INNER  
        DISABLED
        : 
        LEFT
      </span>
      <span
        data-trace="
backgroundColor @inner-component.sheet.root[0]
color @inner-component.sheet.root[1]/ifelse-true
borderColor @outer-component.sheet.Right/inner-component.root"
        style="background-color: yellow; color: red; border-color: green;"
      >
        INNER  
        ENABLED
        : 
        RIGHT
      </span>
    </div>
  </div>
</div>
`;

exports[`STYLED COMPONENT ## WEB ## 01: Outer 1`] = `
<div>
  <div
    class="b"
    data-testid="toggle"
    data-trace="
.b { padding:10px } @outer-component.sheet.root/ifelse-false"
  >
    OUTER 
    CLOSED
    : 
    <span
      class="e c i g"
      data-trace="
.e { font-weight:300 } @outer-component.sheet.Left/inner-component.root[1]/ifelse-false
.c { border-color:red } @outer-component.sheet.Left/inner-component.root[0]
.i { color:gray } @inner-component.sheet.root[1]/ifelse-false
.g { background-color:yellow } @inner-component.sheet.root[0]"
    >
      INNER  
      DISABLED
      : 
      LEFT
    </span>
    <span
      class="f h g"
      data-trace="
.f { border-color:green } @outer-component.sheet.Right/inner-component.root
.h { color:red } @inner-component.sheet.root[1]/ifelse-true
.g { background-color:yellow } @inner-component.sheet.root[0]"
    >
      INNER  
      ENABLED
      : 
      RIGHT
    </span>
  </div>
</div>
`;

exports[`STYLED COMPONENT ## WEB ## 02: Outer opened 1`] = `
<div>
  <div
    class="a"
    data-testid="toggle"
    data-trace="
.a { padding:0 } @outer-component.sheet.root/ifelse-true"
  >
    OUTER 
    OPENED
    : 
    <span
      class="d c i g"
      data-trace="
.d { font-weight:500 } @outer-component.sheet.Left/inner-component.root[1]/ifelse-true
.c { border-color:red } @outer-component.sheet.Left/inner-component.root[0]
.i { color:gray } @inner-component.sheet.root[1]/ifelse-false
.g { background-color:yellow } @inner-component.sheet.root[0]"
    >
      INNER  
      DISABLED
      : 
      LEFT
    </span>
    <span
      class="f h g"
      data-trace="
.f { border-color:green } @outer-component.sheet.Right/inner-component.root
.h { color:red } @inner-component.sheet.root[1]/ifelse-true
.g { background-color:yellow } @inner-component.sheet.root[0]"
    >
      INNER  
      ENABLED
      : 
      RIGHT
    </span>
  </div>
</div>
`;

exports[`STYLED COMPONENT ## WEB ## 03: Outer both 1`] = `
<div>
  <div>
    <div
      class="a"
      data-testid="toggle"
      data-trace="
.a { padding:0 } @outer-component.sheet.root/ifelse-true"
    >
      OUTER 
      OPENED
      : 
      <span
        class="d c i g"
        data-trace="
.d { font-weight:500 } @outer-component.sheet.Left/inner-component.root[1]/ifelse-true
.c { border-color:red } @outer-component.sheet.Left/inner-component.root[0]
.i { color:gray } @inner-component.sheet.root[1]/ifelse-false
.g { background-color:yellow } @inner-component.sheet.root[0]"
      >
        INNER  
        DISABLED
        : 
        LEFT
      </span>
      <span
        class="f h g"
        data-trace="
.f { border-color:green } @outer-component.sheet.Right/inner-component.root
.h { color:red } @inner-component.sheet.root[1]/ifelse-true
.g { background-color:yellow } @inner-component.sheet.root[0]"
      >
        INNER  
        ENABLED
        : 
        RIGHT
      </span>
    </div>
    <div
      class="b"
      data-testid="toggle"
      data-trace="
.b { padding:10px } @outer-component.sheet.root/ifelse-false"
    >
      OUTER 
      CLOSED
      : 
      <span
        class="e c i g"
        data-trace="
.e { font-weight:300 } @outer-component.sheet.Left/inner-component.root[1]/ifelse-false
.c { border-color:red } @outer-component.sheet.Left/inner-component.root[0]
.i { color:gray } @inner-component.sheet.root[1]/ifelse-false
.g { background-color:yellow } @inner-component.sheet.root[0]"
      >
        INNER  
        DISABLED
        : 
        LEFT
      </span>
      <span
        class="f h g"
        data-trace="
.f { border-color:green } @outer-component.sheet.Right/inner-component.root
.h { color:red } @inner-component.sheet.root[1]/ifelse-true
.g { background-color:yellow } @inner-component.sheet.root[0]"
      >
        INNER  
        ENABLED
        : 
        RIGHT
      </span>
    </div>
  </div>
</div>
`;

exports[`STYLED COMPONENT ## WEB ## 04: change open with click 1`] = `
<div>
  <div>
    <div
      class="b"
      data-testid="toggle"
      data-trace="
.b { padding:10px } @outer-component.sheet.root/ifelse-false"
    >
      OUTER 
      CLOSED
      : 
      <span
        class="e c i g"
        data-trace="
.e { font-weight:300 } @outer-component.sheet.Left/inner-component.root[1]/ifelse-false
.c { border-color:red } @outer-component.sheet.Left/inner-component.root[0]
.i { color:gray } @inner-component.sheet.root[1]/ifelse-false
.g { background-color:yellow } @inner-component.sheet.root[0]"
      >
        INNER  
        DISABLED
        : 
        LEFT
      </span>
      <span
        class="f h g"
        data-trace="
.f { border-color:green } @outer-component.sheet.Right/inner-component.root
.h { color:red } @inner-component.sheet.root[1]/ifelse-true
.g { background-color:yellow } @inner-component.sheet.root[0]"
      >
        INNER  
        ENABLED
        : 
        RIGHT
      </span>
    </div>
  </div>
</div>
`;

exports[`STYLED COMPONENT ## WEB ## 04: change open with click 2`] = `
<div>
  <div>
    <div
      class="a"
      data-testid="toggle"
      data-trace="
.a { padding:0 } @outer-component.sheet.root/ifelse-true"
    >
      OUTER 
      OPENED
      : 
      <span
        class="d c i g"
        data-trace="
.d { font-weight:500 } @outer-component.sheet.Left/inner-component.root[1]/ifelse-true
.c { border-color:red } @outer-component.sheet.Left/inner-component.root[0]
.i { color:gray } @inner-component.sheet.root[1]/ifelse-false
.g { background-color:yellow } @inner-component.sheet.root[0]"
      >
        INNER  
        DISABLED
        : 
        LEFT
      </span>
      <span
        class="f h g"
        data-trace="
.f { border-color:green } @outer-component.sheet.Right/inner-component.root
.h { color:red } @inner-component.sheet.root[1]/ifelse-true
.g { background-color:yellow } @inner-component.sheet.root[0]"
      >
        INNER  
        ENABLED
        : 
        RIGHT
      </span>
    </div>
  </div>
</div>
`;
